### PR TITLE
fix error from betabinom stats using only integers for a and b

### LINKED ISSUE #18026: BUG: stats: Error from e.g. `stats.betabinom.stats(10, 2, 3, moments='k')`
When integers are given as the `a` and `b` parameters in a call of `stats.betabinom.stats(n, a, b, moments='k')`, the function raises an exception.

For example, 

```
In [1]: from scipy.stats import betabinom

In [2]: betabinom.stats(10, 2, 3, moments='k')
---------------------------------------------------------------------------
UFuncTypeError                            Traceback (most recent call last)
Cell In [2], line 1
----> 1 betabinom.stats(10, 2, 3, moments='k')

File ~/repos/git/forks/scipy/build-install/lib/python3.10/site-packages/scipy/stats/_distn_infrastructure.py:1124, in rv_generic.stats(self, *args, **kwds)
   1121 scale, loc, goodargs = goodargs[-2], goodargs[-1], goodargs[:-2]
   1123 if self._stats_has_moments:
-> 1124     mu, mu2, g1, g2 = self._stats(*goodargs,
   1125                                   **{'moments': moments})
   1126 else:
   1127     mu, mu2, g1, g2 = self._stats(*goodargs)

File ~/repos/git/forks/scipy/build-install/lib/python3.10/site-packages/scipy/stats/_discrete_distns.py:254, in betabinom_gen._stats(self, n, a, b, moments)
    252 g2 += 3 * a * b * (n - 2)
    253 g2 += 6 * n ** 2
--> 254 g2 -= 3 * e_p * b * n * (6 - n)
    255 g2 -= 18 * e_p * e_q * n ** 2
    256 g2 *= (a + b) ** 2 * (1 + a + b)

UFuncTypeError: Cannot cast ufunc 'subtract' output from dtype('float64') to dtype('int64') with casting rule 'same_kind'
```
If either `a` or `b` is a float, the error does not occur:
```
In [3]: betabinom.stats(10, 2, 3.0, moments='k')
Out[3]: -0.6904761904761907

```
The above was run with the main development branch of SciPy (`'1.11.0.dev0+1559.890ffe9'`) and NumPy 1.24.2.

### SciPy/NumPy/Python version information

/home/warren/repos/git/forks/scipy/build-install/lib/python3.10/site-packages/scipy/__config__.py:140: UserWarning: Install `pyyaml` for better output   warnings.warn("Install `pyyaml` for better output", stacklevel=1) {   "Compilers": {     "c": {       "name": "gcc",       "linker": "ld.bfd",       "version": "11.3.0",       "commands": "cc"     },     "cython": {       "name": "cython",       "linker": "cython",       "version": "0.29.33",       "commands": "cython"     },     "c++": {       "name": "gcc",       "linker": "ld.bfd",       "version": "11.3.0",       "commands": "c++"     },     "fortran": {       "name": "gcc",       "linker": "ld.bfd",       "version": "11.3.0",       "commands": "gfortran"     },     "pythran": {       "version": "0.12.1",       "include directory": "/home/warren/py3.10.8/lib/python3.10/site-packages/pythran"     }   },   "Machine Information": {     "host": {       "cpu": "x86_64",       "family": "x86_64",       "endian": "little",       "system": "linux"     },     "build": {       "cpu": "x86_64",       "family": "x86_64",       "endian": "little",       "system": "linux"     },     "cross-compiled": false   },   "Build Dependencies": {     "blas": {       "name": "openblas",       "found": true,       "version": "0.3.20",       "detection method": "pkgconfig",       "include directory": "/usr/include/x86_64-linux-gnu/openblas-pthread/",       "lib directory": "/usr/lib/x86_64-linux-gnu/openblas-pthread/",       "openblas configuration": "USE_64BITINT= DYNAMIC_ARCH=1 DYNAMIC_OLDER=1 NO_CBLAS= NO_LAPACK= NO_LAPACKE=1 NO_AFFINITY=1 USE_OPENMP=0 generic MAX_THREADS=64",       "pc file directory": "/usr/lib/x86_64-linux-gnu/pkgconfig"     },     "lapack": {       "name": "openblas",       "found": true,       "version": "0.3.20",       "detection method": "pkgconfig",       "include directory": "/usr/include/x86_64-linux-gnu/openblas-pthread/",       "lib directory": "/usr/lib/x86_64-linux-gnu/openblas-pthread/",       "openblas configuration": "USE_64BITINT= DYNAMIC_ARCH=1 DYNAMIC_OLDER=1 NO_CBLAS= NO_LAPACK= NO_LAPACKE=1 NO_AFFINITY=1 USE_OPENMP=0 generic MAX_THREADS=64",       "pc file directory": "/usr/lib/x86_64-linux-gnu/pkgconfig"     }   },   "Python Information": {     "path": "/home/warren/py3.10.8/bin/python3",     "version": "3.10"   } }

### PR DESCRIPTION
**Pull request by vmpyr**
#### Reference issue

Closes: #18026 

The issue stemmed from the fact that we were trying to substract a `float` from an `int64` array. This does not work with operators like `+=` or `-=`. Hence, we need to use `-` and `+` "normally".

Alternatively, I could also have explicitly cast the variable as `float` but this felt more sensible to me.

### CODE REVIEW

**Comment by mdhaber** [comment]
Would it be sufficient to change one line (wherever the result might first be a floating point array) instead of all of them? Where possible, mutating the original array rather than creating copies may be preferable for memory and performance.

This also needs a unit test to confirm that the issue is resolved.

**Comment by vmpyr** [comment]
Okay, let me try the changes you mentioned.

>Please hold off on submitting new PRs while the old ones are reviewed. This will avoid the need to make the same sorts of changes on each one.

Sorry, I did not get you here

**Comment by vmpyr** [comment]
>Please wait before submitting new PRs while the your other PRs are reviewed. This will avoid the need to make the same sorts of changes on each one.

I don't get how that is a problem. I create a new branch for every new PR off of the main upstream branch. Finally everything will be merged in the same codebase. Other places where I contribute, this is fine. Does scipy deploy some other strategy?

**Comment by mdhaber** [comment]
It is not a policy of SciPy; it is a request from me if you'd like me to review your PRs. It is not a question of merge conflicts but reviewer effort. Since these are your first contributions to the library, there will likely be some things to fix. More often than not, the _same_ sorts of things will need to be fixed about several PRs. If you open one PR and learn from the review process before opening the next PR, you are less likely to need to fix the same shortcomings in future PRs, and reviewers will not need to make the same comments multiple times.

**Comment by vmpyr** [comment]
I understand now. I think you took my previous comment to be in a rude tone. It never intended it to be so. I will keep your suggestions in mind for future contributions.
>But if you'd prefer, I don't need to be the one to review the PRs.

Rest assured, I have no preference for anyone reviewing my PRs. I am more than glad with you reviewing them.

**Comment by mdhaber** [comment]
It looks like this still fails:
```python3
betabinom.stats(10., 2, 3, moments='k')  # n is a float now
```

**Review by mdhaber — COMMENTED** [review]
Some other minor points. I'll push the fix.

File: scipy/stats/tests/test_discrete_distns.py [hunk 1]
```diff
@@ -141,6 +143,13 @@ def test_betabinom_a_and_b_unity():
     expected = np.repeat(1 / (n + 1), n + 1)
     assert_almost_equal(p, expected)
 
+@pytest.mark.parametrize('dtypes', itertools.product(*[(int, float)]*3))
```
Comment (mdhaber) [inline scipy/stats/tests/test_discrete_distns.py:146]: Oops, last thing: PEP8 requires two empty lines between top level function definitions.
```suggestion

@pytest.mark.parametrize('dtypes', itertools.product(*[(int, float)]*3))
```

File: scipy/stats/tests/test_discrete_distns.py [hunk 2]
```diff
@@ -141,6 +143,13 @@ def test_betabinom_a_and_b_unity():
...
+def test_betabinom_stats_a_and_b_integers_gh18026(dtypes):
+    # gh-18026 reported that `betabinom` kurtosis calculation fails when some
+    # parameters are integers. Check that this is resolved.
+    n_type, a_type, b_type = dtypes
+    n, a, b = n_type(10), a_type(2), b_type(3)
+    assert_allclose(betabinom.stats(n, a, b, moments='k'), -0.6904761904761907)
```
Comment (mdhaber) [inline scipy/stats/tests/test_discrete_distns.py:152]:
```suggestion
    assert_allclose(betabinom.stats(n, a, b, moments='k'), -0.6904761904761907)

```

**Review by mdhaber — APPROVED** [review]

**Comment by mdhaber** [comment]
Thanks @vmpyr. I think your other PRs will be reviewed again shortly.

**Comment by vmpyr** [comment]
>Some other minor points. I'll push the fix.

Thank you, I am busy with exams currently so didn't see this earlier. 

Thank you @mdhaber for the timely and insightful reviews. I'll resume work on other PRs asap.

**Pull request merged**